### DISCUSSION TITLE
Allowing filters to be asynchronous

### DISCUSSION THREAD

**Discussion by slaymaker1907** [TiddlyWiki/TiddlyWiki5#6550] [Ideas]
I have a branch on my fork that makes filters and filterrunprefixes (since it was necessary to implement async filters) async. The synchronous functions for filtering also exist to avoid compatibility issues, but I think it is an idea worth considering. It would allow for doing things like REST calls in filters much more easily and would help prevent long running/infinite filters from crashing a wiki.

One filter in particular I'd like an async version for would be regular expressions. These can easily crash wikis if they are poorly written and this problem could be alleviated by running the regex filter asynchronously in a WebWorker (probably also keeping a copy of the wiki in the worker so it can be done in a batched way).

https://github.com/slaymaker1907/TiddlyWiki5/tree/support-async-filters

It's a bit tedious writing this async code without promises, but I wanted to avoid the question of whether to polyfill/drop support for IE. Callbacks also tend to have slightly better performance. The code does use promises when available to reset the call stack without the overhead of setTimeout (which when nested can result in a 4-5ms delay per use of setTimeout).

The branch also implements some changes to Widget.refresh() to support asynchronicity, but the commits are divided up so 5e6e860c68d77a362a7ee63ec4a036e700cc3584 contains just the changes to support async filtering.

**Comment by slaymaker1907** [comment]
The current standard I've been using to make async APIs is (syncArgs...,callback) so callback is the last parameter. Each callback follows the node standard where the first parameter is err and the remaining ones are the rest of the results.

For refresh(), I modified the method to accept a second optional parameter "getCallback". It just returns a callback but allows the caller of refresh() to see whether the object supports async refresh or not. Would this break any known existing widgets @Jermolene ? I saw one weird case where refresh() was called with multiple parameters aside from the changed tiddlers.

**Comment by slaymaker1907** [comment]
Right now, the code just assumes that all filterrunprefix modules can be made to support asynchronicity. However, it could also be done in a backwards compatible way, but it would require checking the input filter to see if it uses any prefixes that don't support async. Unlike filters, it isn't possible to convert a sync prefix to be async.

**Comment by linonetwo** [comment]
When dealing with async code, it is important to make sure error stack is easy to read. Have you tested the case when filter throw an error in it?

And Promise and async function is better in this than callback and setTimeout.

**Reply by pmario** [reply]
>And Promise and async function is better in this than callback and setTimeout.

TW core code only allows ES5 at the moment. But there are some discussions about ESxx at: **FYI: IE to finally retire June 2022** #5875 and  **Be explicit about browser support targets** #2278

**Reply by linonetwo** [reply]
Thanks, I will take a look

**Reply by slaymaker1907** [reply]
I agree, I was just hesitant to add in promise code since Tiddlywiki doesn't currently use them. In terms of error stacks, I haven't tested it and I know things can definitely go wrong since I'm not wrapping things in try/catch (so errors might happen and never be propagated upward). That is one of the things that would need to be done before this could ever be part of a release.

**Comment by Jermolene** [comment]
Hi @slaymaker1907 and apologies for missing this and hence the very late reply.

> I have a branch on my fork that makes filters and filterrunprefixes (since it was necessary to implement async filters) async. The synchronous functions for filtering also exist to avoid compatibility issues, but I think it is an idea worth considering.

TiddlyWiki makes careful use of JavaScript's non-preemptive event driven multitasking model. The refresh process is delayed to an async timer-based task that must run to completion before allowing other tasks to run. We couldn't make filters be asynchronous without also making the refresh process asynchronous, which would break the current model.

> One filter in particular I'd like an async version for would be regular expressions. These can easily crash wikis if they are poorly written and this problem could be alleviated by running the regex filter asynchronously in a WebWorker (probably also keeping a copy of the wiki in the worker so it can be done in a batched way).

The recommended approach for integrating asynchronous tasks is to trigger them via a message to the root widget, and have them asynchronously modify the tiddler store to communicate results.

**Reply by linonetwo** [reply]
In my daily job, We tried to refactor a low code engine's rendering logic from sync to be async, and found it very hard to change every aspect of the engine and test them, finally, we only decided to make a separate async engine prototype instead. So I agree this will need a huge amount of effort.

**Reply by slaymaker1907** [reply]
@Jermolene modifying the store asynchronously is a good idea. I noticed the issue with synchronous refresh which is why I had the idea of having an optional async refresh. However, that would only work if the entire path to the widget starting from root supports async.

That gives me an idea for how to make the list widget support async in a nicer way. It could just work as if the filter is empty (maybe transclude a loading sign or something) then pretend as if the store was changed to force a refresh once the data is ready. However, even with that design, async filters would still be useful (or at least a way to run them non-blocking such as running the filter operations in a webworker).

**Reply by Jermolene** [reply]
Hi @slaymaker1907 

> I noticed the issue with synchronous refresh which is why I had the idea of having an optional async refresh. However, that would only work if the entire path to the widget starting from root supports async.

It is correct that it is always possible to wrap a sync function to make it async, and thus from a coding point of view it would be possible to make the refresh process async. The trouble that then arises is that that would break the threading assumptions that underpin the core architecture.

> That gives me an idea for how to make the list widget support async in a nicer way. It could just work as if the filter is empty (maybe transclude a loading sign or something) then pretend as if the store was changed to force a refresh once the data is ready. However, even with that design, async filters would still be useful (or at least a way to run them non-blocking such as running the filter operations in a webworker).

You'll run into the same problem again, I'm afraid, and be looking at rearchitecting everything.

But, as I say, all of the use cases you mentioned are already supported by the TW core, just not quite in the way that you've envisaged.

For example, to run a filter in a web worker one would use the message passing mechanism to kick off a `postMessage()` to the worker thread. The worker thread would then asynchronously return the results when it is ready. The UI then uses an ordinary synchronous filter that catches the tiddlers that are expected in return from the worker. Each time a tiddler comes in a refresh will be triggered and the UI will be updated to incorporate the updates.

**Reply by linonetwo** [reply]
I think this worth a try, to fix the performance issue in https://github.com/Jermolene/TiddlyWiki5/issues/7151 

But we have to do this in the core, so the user side can transparently get the performance boost. Just modify the core filter execution code, if is in a client-server setup, send it to server instead, and trigger rerender on return.

**Reply by Jermolene** [reply]
Hi @linonetwo the trouble is once again that the filter execution method is synchronous, so it is not possible to wait for results from the server before returning.

**Reply by linonetwo** [reply]
Instead of run some async operators, we can make `wiki.filterTiddlers` return promise, and inside it just fetch the nodejs server API.

But this will require `ListWidget.prototype.getTiddlerList` return promise (with an option to enable this), and `ListWidget.prototype.handleListChanges` have to be async too, but we will stop here, because the call to `this.insertListItem` can be sync after we await the async data.

This will need to change many core widgets, to add await...3rd party plugin widgets didn't pass new option won't receive promise, so won't be a breaking change. And we can enable this feature in core widgets one by one, so slowly achieve this.

Just a thought, this will only happened in next year, if it is me that are going to write it...

**Reply by Jermolene** [reply]
Hi @linonetwo as I tried to explain above, asynchronous code cannot be swapped in to replace synchronous functionality. Making filters and widgets operate asynchronously would require a fundamental redesign of how TW handles refreshing. Right now the fact that the refresh cycle is single threaded is an essential part of the architecture.

**Reply by slaymaker1907** [reply]
@Jermolene and @linonetwo, I have an idea of how to make this work while avoiding fundamental changes to the refresh process. I know @Jermolene mentioned using the store to do async stuff, but I think we could instead use the optional async filter functions in ListWidget.

How it would work is that the ListWidget would update immediately by displaying blank content (and ignoring any child widgets/actions) until the async function actually returns. The blank would probably be customizable as a field taking a tiddler to transclude while loading. I think most of the issues with poor performance come directly or indirectly from ListWidget since it's the one that typically deals with arbitrary user filters.

The one aspect that I'm not sure about is if this change would require a state tiddler or something to communicate to TiddlyWiki when the results are complete and it needs to rerender. I'm thinking that we could have a global tiddler that just tracks a count for the number of ListWidgets which have completed an async request (therefore, when a request completes, we can just increment that counter which will force a render).

**Reply by Jermolene** [reply]
Hi @slaymaker1907 it's an interesting approach, but I don't think it necessarily needs so much reengineering.

TiddlyWiki is designed to support asynchronous queries – you can see the pattern in use in the plugin library implementation in control panel.

The basic idea is to use a message to trigger an asynchronous query that asynchronously creates tiddlers as the data becomes available. TiddlyWiki's standard refresh mechanism ensures that the UI is updated to reflect the arrival of those tiddlers.

**Comment by linonetwo** [comment]
Maybe can solve #5463 running regex in background.

What I think is more valuable of async filter is, this may enables us run filters on nodejs side or on a remote db side, so website like tiddlylinks can be truely dynamic, loads content from the backend, like reddit. and #5568

**Reply by linonetwo** [reply]
But as Jeremy said, we always have existed way to do it, just maybe need to write more complicated code to implement.

So though I'm also interested in loading async data, I will spend time on other priorities at the moment... like the wysiwyg editor...
